services/comps/block/src/request_queue.rs

```diff
@@ -7,24 +7,113 @@ use super::{
     id::Sid,
 };
 
-/// Represents the software staging queue for the `BioRequest` objects.
-pub trait BioRequestQueue {
+use aster_frame::sync::{Mutex, WaitQueue};
+
+/// A simple block I/O request queue with a single queue.
+///
+/// It is a FIFO producer-consumer queue, where the producer (e.g., filesystem)
+/// submits requests to the queue, and the consumer (e.g., block device driver)
+/// continuously consumes and processes these requests from the queue.
+pub struct BioRequestSingleQueue {
+    queue: Mutex<VecDeque<BioRequest>>,
+    num_requests: AtomicUsize,
+    wait_queue: WaitQueue,
+}
+
+impl BioRequestSingleQueue {
+    /// Creates an empty queue.
+    pub fn new() -> Self {
+        Self {
+            queue: Mutex::new(VecDeque::new()),
+            num_requests: AtomicUsize::new(0),
+            wait_queue: WaitQueue::new(),
+        }
+    }
+
+    /// Returns the number of requests currently in this queue.
+    pub fn num_requests(&self) -> usize {
+        self.num_requests.load(Ordering::Relaxed)
+    }
+
     /// Enqueues a `SubmittedBio` to this queue.
     ///
-    /// This `SubmittedBio` will be merged into an existing `BioRequest`, or a new
-    /// `BioRequest` will be created from the `SubmittedBio` before being placed
-    /// into the queue.
+    /// When enqueueing the `SubmittedBio`, try to insert it into the last request if the
+    /// type is same and the sector range is contiguous.
+    /// Otherwise, creates and inserts a new request for the `SubmittedBio`.
     ///
     /// This method will wake up the waiter if a new `BioRequest` is enqueued.
-    fn enqueue(&self, bio: SubmittedBio) -> Result<(), BioEnqueueError>;
+    pub fn enqueue(&self, bio: SubmittedBio) -> Result<(), BioEnqueueError> {
+        let mut queue = self.queue.lock();
+        if let Some(request) = queue.front_mut() {
+            if request.can_merge(&bio) {
+                request.merge_bio(bio);
+                return Ok(());
+            }
+        }
+
+        let new_request = BioRequest::from(bio);
+        queue.push_front(new_request);
+        drop(queue);
+        self.inc_num_requests();
```

Comment:
In certain race conditions between concurrent `enqueue` and `dequeue`, the counter may be decreased first, making the number underflow temporarily. This should be avoided.

Reply:
Yes the modification of `num_requests` should be controlled by the lock of  queue.